apps/ewallet/lib/ewallet/gates/token_gate.ex

```diff
@@ -0,0 +1,131 @@
+# Copyright 2018-2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule EWallet.TokenGate do
+  @moduledoc false
+
+  alias EWallet.BlockchainHelper
+  alias EWalletDB.{BlockchainWallet, Token}
+
+  @doc """
+  Validate that the `decimals` and `symbol` of the token are the same as
+  the ones defined in the erc20 contract. If the contract does not implement
+  these fields, we rely on the token's field values.
+  Returns {:ok, status} where `status` is the blockchain status of the token.
+  The status is "confirmed" if the hot wallet balance is positive, or "pending" otherwise.
+  """
+  def validate_erc20_readiness(contract_address, %{
+        symbol: symbol,
+        subunit_to_unit: subunit_to_unit
+      }) do
+    with {:ok, erc20_attrs} <- get_erc20_capabilities(contract_address),
+         :ok <- validate_decimals(erc20_attrs, subunit_to_unit),
+         :ok <- validate_symbol(erc20_attrs, symbol) do
+      {:ok, get_blockchain_status(erc20_attrs)}
+    else
+      :error -> {:error, :token_not_matching_contract_info}
+      error -> error
+    end
+  end
+
+  defp validate_decimals(%{decimals: value}, subunit_to_unit) do
+    case value == :math.log10(subunit_to_unit) do
+      true -> :ok
+      false -> :error
+    end
+  end
+
+  defp validate_decimals(_, _), do: :ok
```

Comment:
Should we actually let this case pass? I think we can remove it and let exception does its job if it's not expected to happen.

Reply:
Will add more doc to make it clear what each case handle